tools/clang/test/HLSLFileCheck/hlsl/intrinsics/mul/mul.hlsl

```diff
@@ -0,0 +1,992 @@
+// IN1_MAT2_OVRLD_OD: extractvalue %dx.types.CBufRet.i32
+// IN1_MAT2_OVRLD_OD: extractvalue %dx.types.CBufRet.i32
+// IN1_MAT2_OVRLD_OD: extractvalue %dx.types.CBufRet.i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+// IN1_MAT2_OVRLD_OD: mul i32
+
+// mixed: vector and matrix
+// RUN: %dxc -T vs_6_0 -E main -DELEM_TY1=int2 -DELEM_TY2=int2x4 -DRET_TY=int4 %s  | FileCheck %s -check-prefix=IN2_MAT1_OVRLD
+// IN2_MAT1_OVRLD: mul i32
+// IN2_MAT1_OVRLD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD: mul i32
+// IN2_MAT1_OVRLD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD: mul i32
+// IN2_MAT1_OVRLD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD: mul i32
+
+// RUN: %dxc -T vs_6_0 -E main -DELEM_TY1=int4x3 -DELEM_TY2=int3 -DRET_TY=int4 %s  | FileCheck %s -check-prefix=IN3_MAT2_OVRLD
+// IN3_MAT2_OVRLD: mul i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: mul i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: mul i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: mul i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD: call i32 @dx.op.tertiary.i32
+
+// RUN: %dxc -T vs_6_0 -E main -Od -DELEM_TY1=int2 -DELEM_TY2=int2x4 -DRET_TY=int4 %s  | FileCheck %s -check-prefix=IN2_MAT1_OVRLD_OD
+// IN2_MAT1_OVRLD_OD: mul i32
+// IN2_MAT1_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD_OD: mul i32
+// IN2_MAT1_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD_OD: mul i32
+// IN2_MAT1_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN2_MAT1_OVRLD_OD: mul i32
+
+// RUN: %dxc -T vs_6_0 -E main -Od -DELEM_TY1=int4x3 -DELEM_TY2=int3 -DRET_TY=int4 %s  | FileCheck %s -check-prefix=IN3_MAT2_OVRLD_OD
+// IN3_MAT2_OVRLD_OD: mul i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: mul i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: mul i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: mul i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+// IN3_MAT2_OVRLD_OD: call i32 @dx.op.tertiary.i32
+
+// *****************************
+// uint overloads
+// *****************************
+
+// vectors and scalars
+// RUN: %dxc -T vs_6_0 -E main -DELEM_TY1=uint4 -DELEM_TY2=uint4 -DRET_TY=uint %s  | FileCheck %s -check-prefix=UIN4_OVRLD
+// UIN4_OVRLD: call i32 @dx.op.tertiary.i32
```

Comment:
Should you be checking the opcode to make sure it's UMAD for these unsigned cases?  Likewise making sure it's IMAD for the signed ones, and FMAD for float cases?

Reply:
Done.

While updating the tests to also check for opcode, I found a new bug which exists in master. Matrix multiply of unsigned ints, uses `IMad` instead of `UMad`. Issue filed #2482 and added a reference in the TODO.